Block 1: Import libraries

In [6]:
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn import metrics

Block 2: Function

* Have an input argument into the function that determines whether you want plots or not
* If you do want a figure, it should be a subplot (2,1) that shows a box plot of (y , X) and the ROC curve. Make sure the plots are labeled and legible so that you can show Pat the plots

In [7]:
def survival(X, y, plot):
  
  # decision tree binary classification with train and test split

  Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, random_state=1)

  tree_model = tree.DecisionTreeClassifier(max_depth=1)
  tree_model.fit(Xtrain, ytrain)

  training_score = tree_model.score(Xtrain, ytrain)
  testing_score = tree_model.score(Xtest, ytest)

  X = titanic['pclass']
  if plot == 'yes':
    fig, axes = plt.subplots(2, 1)
    sns.boxplot(y, X, ax=axes[0])
    ypred = np.where(tree_model.predict(Xtest), 1, 0)
    fpr_tree,tpr_tree,thresholds_tree = metrics.roc_curve(ytest, ypred, pos_label=1)
    print(fpr_tree)
    print(tpr_tree)
    axes[1].plot(fpr_tree,tpr_tree,label='tree')
    axes[1].plot([0,1],[0,1],color=[0,0,0],linestyle=':', label='chance')
    axes[1].set_xlabel('false positive rate')
    axes[1].set_ylabel('true positive rate')
    axes[1].legend()
    fig.tight_layout()

  return training_score, testing_score

Block 3: Load data

In [12]:
titanic = sns.load_dataset('titanic')
titanic.head()

survived  pclass     sex   age  ...  deck  embark_town  alive  alone
0         0       3    male  22.0  ...   NaN  Southampton     no  False
1         1       1  female  38.0  ...     C    Cherbourg    yes  False
2         1       3  female  26.0  ...   NaN  Southampton    yes   True
3         1       1  female  35.0  ...     C  Southampton    yes  False
4         0       3    male  35.0  ...   NaN  Southampton     no   True

[5 rows x 15 columns]

Block 4: Use function with pclass

In [13]:
survival(titanic[['pclass']], titanic['alive'], 'no')

(0.6751497005988024, 0.6457399103139013)

Block 5: Use function with fare

In [10]:
survival(titanic[['fare']], titanic['alive'], 'no')

(0.6302395209580839, 0.5739910313901345)

The testing score for pclass is 0.64, which is higher than the testing score for fare at 0.57.

Pclass is a better predictor of survival, so Pat should focus on being in a better class.